src/chunkserver/output_buffer.cc

```diff
@@ -30,71 +30,119 @@
 #include "common/crc.h"
 #include "common/massert.h"
 
-OutputBuffer::OutputBuffer(size_t internalBufferCapacity)
-    : internalBufferCapacity_(internalBufferCapacity),
-      internalBufferCapacityAligned_(getAlignedSize(internalBufferCapacity_)),
-      padding_(internalBufferCapacityAligned_ - internalBufferCapacity_),
-      buffer_(internalBufferCapacityAligned_, 0),
-      bufferUnflushedDataFirstIndex_(padding_),
-      bufferUnflushedDataOneAfterLastIndex_(padding_) {
-	eassert(internalBufferCapacity > 0);
-	buffer_.reserve(internalBufferCapacityAligned_);
-}
+OutputBuffer::OutputBuffer(size_t headerSize, size_t numBlocks)
+    : currentRemainingBytesForFD_(0),
+      headerSize_(headerSize),
+      numBlocks_(numBlocks),
+      blockBuffer_(numBlocks * SFSBLOCKSIZE, disk::kIoBlockSize),
+      crcBuffer_(numBlocks * kCrcSize),
+      headerBuffer_(numBlocks * headerSize) {}
 
 OutputBuffer::WriteStatus OutputBuffer::writeOutToAFileDescriptor(int outputFileDescriptor) {
+	// Let's write block by block
 	while (bytesInABuffer() > 0) {
-		ssize_t ret = ::write(outputFileDescriptor, &buffer_[bufferUnflushedDataFirstIndex_],
-				bytesInABuffer());
+		if (currentRemainingBytesForFD_ == 0) {
+			// Prepare the blockBuffer to write the current block:
+			// - move back the unflushed data in block buffer
+			// - copy the header
+			// - copy the crc
+			// - set the currentRemainingBytesForFD_
+			blockBuffer_.moveUnflushedDataFirstIndex(-(int32_t)(kCrcSize));
+			crcBuffer_.copyFromBuffer(blockBuffer_.getUnflushedDataFirstIndex(), kCrcSize);
+			blockBuffer_.moveUnflushedDataFirstIndex(-(int32_t)(headerSize_));
+			headerBuffer_.copyFromBuffer(blockBuffer_.getUnflushedDataFirstIndex(), headerSize_);
+			currentRemainingBytesForFD_ =
+			    std::min(SFSBLOCKSIZE + kCrcSize + headerSize_, bytesInABuffer());
+		}
+		ssize_t ret = ::write(outputFileDescriptor, blockBuffer_.getUnflushedDataFirstIndex(),
+		                      currentRemainingBytesForFD_);
+
 		if (ret <= 0) {
-			if (ret == 0 || errno == EAGAIN) {
-				return WRITE_AGAIN;
-			}
-			return WRITE_ERROR;
+			if (ret == 0 || errno == EAGAIN) { return WriteStatus::Again; }
```

Comment:
Consider this change here:
```suggestion
			if (!ret || errno == EAGAIN) { return WriteStatus::Again; }
```

Reply:
I'd rather be explicit and check for 0 explicitly, even though negatives convert to true. Seems more clear to me.

I don't know if it's logged above, but even then I think you should log errno here (as an error), to better trace it later if needed.

Reply:
Of course, if it's logged above, you probably need to remove or change the errno checking there, from man 3 errno:

```
      A common mistake is to do

           if (somecall() == -1) {
               printf("somecall() failed\n");
               if (errno == ...) { ... }
           }

       where errno no longer needs to have the value it had  upon  return  from
       somecall()  (i.e.,  it  may have been changed by the printf(3)).  If the
       value of errno should be preserved across a library  call,  it  must  be
       saved:

           if (somecall() == -1) {
               int errsv = errno;
               printf("somecall() failed\n");
               if (errsv == ...) { ... }
           }
```

Reply:
I agree with Urmas, I think it is better the explicit equal check with 0.

Reply:
Regarding @uristdwarf suggestion of logging errno here: the function which calls this `writeOutToAFileDescriptor` function logs an error using old LOG_NOTICE but does not specify. I could log it here, but then I'll have two logs about the same thing, and modifying it in an above function could already not have the initial errno.

So, I'll opt for leaving it like it is now.